Source/DafnyLanguageServer/Language/Symbols/ScopeSymbol.cs

```diff
@@ -1,15 +1,67 @@
 ﻿using System.Collections.Generic;
 using System.Threading;
+using Microsoft.Boogie;
 
 namespace Microsoft.Dafny.LanguageServer.Language.Symbols {
+  public abstract class DeclarationStmtOrExpr {
+    public IToken Tok;
+    public IToken? EndTok;
+    public object Node;
+
+    protected DeclarationStmtOrExpr(IToken tok, IToken endTok, object node) {
+      Tok = tok;
+      EndTok = endTok;
+      Node = node;
+    }
+  }
+
+  public class StmtDeclaration : DeclarationStmtOrExpr {
+    public BlockStmt Declaration;
+
+    public StmtDeclaration(BlockStmt declaration) : base(declaration.Tok, declaration.EndTok, declaration) {
+      Declaration = declaration;
+    }
+  }
+
+  public class ExpressionDeclaration : DeclarationStmtOrExpr {
+    public Expression Declaration;
+    public ExpressionDeclaration(Expression declaration) : base(declaration.tok, Token.NoToken, declaration) {
+      Declaration = declaration;
+    }
+  }
+
+  public class FunctionDeclaration : DeclarationStmtOrExpr {
+    public Function Declaration;
+    public FunctionDeclaration(Function declaration) : base(declaration.tok, Token.NoToken, declaration) {
+      Declaration = declaration;
+    }
+  }
+
+  public class MethodDeclaration : DeclarationStmtOrExpr {
+    public Method Declaration;
+    public MethodDeclaration(Method declaration) : base(declaration.tok, Token.NoToken, declaration) {
+      Declaration = declaration;
+    }
+  }
   public class ScopeSymbol : Symbol, ILocalizableSymbol {
-    public BlockStmt Declaration { get; }
-    public object Node => Declaration;
+    public DeclarationStmtOrExpr Declaration { get; }
+    public object Node => Declaration.Node;
     public List<ISymbol> Symbols { get; } = new List<ISymbol>();
     public override IEnumerable<ISymbol> Children => Symbols;
 
     public ScopeSymbol(ISymbol? scope, BlockStmt declaration) : base(scope, string.Empty) {
-      Declaration = declaration;
+      Declaration = new StmtDeclaration(declaration);
+    }
+
+    public ScopeSymbol(ISymbol? scope, Expression declaration) : base(scope, string.Empty) {
+      Declaration = new ExpressionDeclaration(declaration);
+    }
+
+    public ScopeSymbol(ISymbol? scope, Function declaration) : base(scope, string.Empty) {
+      Declaration = new FunctionDeclaration(declaration);
+    }
+    public ScopeSymbol(ISymbol? scope, Method declaration) : base(scope, string.Empty) {
+      Declaration = new MethodDeclaration(declaration);
```

Comment:
The only reason why `ScopeSymbol` is aware of different implementations is because of this code snippet in `SymbolTableFactory.cs`
![image](https://user-images.githubusercontent.com/3601079/145597500-982c6a82-3d63-4de3-8aed-1b5caec5d642.png)
Otherwise, just storing the "Node" would be sufficient.
So my question to you is, **why are we registering the location of this ScopeSymbol, if it's anonymous?** Is it for some kind of code folding I never experienced? To me it seems useless as scopesymbols don't have names. Also, Expression-based scopeSymbols don't have end tokens, so that's why I had to add the check.

An `ensures` clause *currently* cannot have all the statements/expressions that the `ScopeSymbol` type would support, but since we are using symbols, this is an acceptable level of abstraction to reuse the same kinds of symbols.
In my implementation, an ensures clause will not create a scope. Only if there are bound variables, then it will create scopes, but attach them to the top-level FunctionSymbol or MethodSymbol. There is no scope for the `ensures` clause like there is for a method body.

Reply:
> So my question to you is, **why are we registering the location of this ScopeSymbol, if it's anonymous?** Is it for some kind of code folding I never experienced? To me it seems useless as scopesymbols don't have names. Also, Expression-based scopeSymbols don't have end tokens, so that's why I had to add the check.

Knowing the location of  `ScopeSymbol` is necessary to correctly handle variable shadowing. See the following example:
![image](https://user-images.githubusercontent.com/18049310/145784957-e6e09acd-6509-4257-abbc-635df3dbbe32.png)
The `x` declared inside the anonymous scope is of type `B`. The language server correctly proposes `b` as a dot-completion suggestion. If it wouldn't know about the anonymous scope, the language server can't decide which of the two `x` is designated here. On the other hand, it proposes `a` outside the anonymous scope.
Moreover, PR #1549 will require this information too.

> In my implementation, an ensures clause will not create a scope. Only if there are bound variables, then it will create scopes, but attach them to the top-level FunctionSymbol or MethodSymbol. There is no scope for the `ensures` clause like there is for a method body.

I guess you're talking about this code:
```cs
      private void ProcessBoundVariableBearingExpression<TExpr>(
        TExpr boundVarExpression, System.Action baseVisit
        ) where TExpr : Expression, IBoundVarsBearingExpression {
        var oldBlock = block;
        if (block.Node != boundVarExpression) { // Else we reuse the current scope
          block = new ScopeSymbol(block, boundVarExpression);
          oldBlock.Symbols.Add(block);
        }
        foreach (var parameter in boundVarExpression.AllBoundVars) {
          block.Symbols.Add(ProcessBoundVar(block, parameter));
        }
        baseVisit();
        block = oldBlock;
      }
```
Could you explain why you're reusing the scope? That would be a better comment here too (we can see what you're doing, not why you're doing it). If it's just an optimization I'd rather go for the simple approach and create a new symbol scope (not necessarily of  type `ScopeSymbol`).

In the meantime I now understand the need for `if (endToken != null && endToken != Token.NoToken)` within `SymbolTableFactory`. To me this appears hacky (sorry for saying so) since the scope/range/location of the expressions is not correctly registered, although, the type `ILocalizableSymbol` intends to depict this.
I see that the `ensures` clause and its siblings are anonymous scopes. However, they should be correctly registered (i.e., the range of the expression) to avoid hard-to-follow behavior in the feature. For example, if we extend the code completion beyond the dot (e.g., support Ctrl+Space), we'll have a hard time with this hack since the anonymous scope is hidden. Honestly, I had a hard time with the debugger, wondering why I couldn't see the `ScopeSymbol` instances until I remembered the if-statement I mentioned before.
Although the code works for the moment, I don't believe it's beneficial in the long-term if we start integrating hacks contradicting the design just for the ease of implementation.

So what I believe to fully implement this, we best split `ScopeSymbol` into two types, for example, `BlockScopeSymbol` (drop-in replacement for `ScopeSymbol` and `ExpressionScopeSymbol` (for the comprehension expressions and ensures/requires/etc. expressions). Or at least simplify `ScopeSymbol` to accept the locations directly to avoid these helper classes. However, if I remember right, I removed this information from the symbol classes to make it evident that the location should be retrieved using `LookupTree` and `Locations` from `SymbolTable`.
Maybe @keyboardDrummer has some valuable input on this discussion as well.

Reply:
I get it better now. So, for auto-completion, it will look for the enclosing scope, and provide variables names accordingly. I wasn't aware of this special interesting feature until you now.

The reason why I'm reusing the scope (and I'll write it down in a comment), is that, when we start creating a scope for a function body, if the first expression is actually a comprehension, it would normally also create a nested scope. However, since the function body is not wrapped, the two scopes now point to the same node in the AST.
When there is the traversal that bind nodes to their respective scope, the the comprehension expression will be bound to the outer scope in the variables "declarations". However, when during a later traversal of the tree, it goes through `ProcessNestedScope` for comprehensions, which assumes that the node has been bound to a scope symbol and will attempt to retrieve it with `declarations[node]`, which in will throw an exception. The integration tests caught that bug.

I added the following shorter comment:
```cs
        // To prevent two scope symbols from pointing to the same node,
        // (this crashes `declarations[nodes]` later on)
        // we reuse the existing scope symbol if it happens to be a top-level
        // bounded variable bearing expression that otherwise would create a new scope symbol
```

For the "hack", you're right, it would be better to have a consistent set of features, so I'll modify the Dafny parser so that it records the final token of the expressions that define scopes so that I can remove the check.
It also helps me to remove all the extra stuff from ScopeSymbol so that now it only stores a new object implementing IRegion, I believe this is in the spirit of the scope symbol.